### DISCUSSION TITLE
how to use rust-tools with mason.nvim?

### DISCUSSION THREAD

**Discussion by kylo252** [LunarVim/LunarVim#3201] [Q&A]
## Description

This completely broke my rust-tools configuration which is 1:1 from https://www.lunarvim.org/languages/rust.html#extra-plugins.

Are those random breaking changes supposed to happen in rolling? Then a more suited branch name would be  'unstable' instead.

_Originally posted by @amoebaguy in https://github.com/LunarVim/LunarVim/issues/2880#issuecomment-1210124209_

## Solution

remove all custom lsp-installer logic, since rust-analyzer is now on path

```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

lvim.plugins = {
  {
   "simrat39/rust-tools.nvim",
    -- ft = { "rust", "rs" }, -- IMPORTANT: re-enabling this seems to break inlay-hints
    config = function()
      require("rust-tools").setup {
        tools = {
          executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
          reload_workspace_from_cargo_toml = true,
          inlay_hints = {
            auto = true,
            only_current_line = false,
            show_parameter_hints = true,
            parameter_hints_prefix = "<-",
            other_hints_prefix = "=>",
            max_len_align = false,
            max_len_align_padding = 1,
            right_align = false,
            right_align_padding = 7,
            highlight = "Comment",
          },
          hover_actions = {
            border = {
              { "╭", "FloatBorder" },
              { "─", "FloatBorder" },
              { "╮", "FloatBorder" },
              { "│", "FloatBorder" },
              { "╯", "FloatBorder" },
              { "─", "FloatBorder" },
              { "╰", "FloatBorder" },
              { "│", "FloatBorder" },
            },
            auto_focus = true,
          },
        },
        server = {
          on_init = require("lvim.lsp").common_on_init,
          on_attach = function(client, bufnr)
            require("lvim.lsp").common_on_attach(client, bufnr)
            local rt = require "rust-tools"
            -- Hover actions
            vim.keymap.set("n", "<C-space>", rt.hover_actions.hover_actions, { buffer = bufnr })
            -- Code action groups
            vim.keymap.set("n", "<leader>lA", rt.code_action_group.code_action_group, { buffer = bufnr })
          end,
        },
      }
    end,
  },
}
```

**Comment by luxus** [comment]
@abzcoding (sorry for pinging) should know something.. he is using rust with mason

**Comment by abzcoding** [comment]
@luxus this works fine for me -> https://github.com/abzcoding/lvim/blob/main/lua/user/rust_tools.lua

**Comment by kylo252** [comment]
> @luxus this works fine for me -> [https://github.com/abzcoding/lvim/blob/main/lua/user/rust_tools.lua](https://github.com/abzcoding/lvim/blob/main/lua/user/rust_tools.lua?rgh-link-date=2022-08-12T11%3A48%3A47Z)

mason provides codelldb, so that config could probably be even simpler 👀

**Comment by rlpowell** [comment]
It took me like hours? to get this working, because I couldn't find a complete-for-newbs thing that covered the whole thing anywhere, so here's where I ended up in case it helps anybody:

```vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

lvim.plugins = {
  {
    "simrat39/rust-tools.nvim",
    config = function()
      local status_ok, rust_tools = pcall(require, "rust-tools")
      if not status_ok then
        return
      end

      local opts = {
        tools = {
          executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
          reload_workspace_from_cargo_toml = true,
          inlay_hints = {
            auto = true,
            only_current_line = false,
            show_parameter_hints = true,
            parameter_hints_prefix = "<-",
            other_hints_prefix = "=>",
            max_len_align = false,
            max_len_align_padding = 1,
            right_align = false,
            right_align_padding = 7,
            highlight = "Comment",
          },
          hover_actions = {
            --border = {
            --        { "╭", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╮", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --        { "╯", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╰", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --},
            auto_focus = true,
          },
        },
        server = {
          on_attach = require("lvim.lsp").common_on_attach,
          on_init = require("lvim.lsp").common_on_init,
          settings = {
            ["rust-analyzer"] = {
              checkOnSave = {
                command = "clippy"
              }
            }
          },
        },
      }
      --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"

      --local codelldb_path = extension_path .. "adapter/codelldb"
      --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"

      --opts.dap = {
      --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
      --}
      rust_tools.setup(opts)
    end,
    ft = { "rust", "rs" },
  },
}
```

I did not try to get debugging working at all, hence things being commented out.

**Reply by datfooldive** [reply]
> It took me like hours? to get this working, because I couldn't find a complete-for-newbs thing that covered the whole thing anywhere, so here's where I ended up in case it helps anybody:
> 
> ```
> 
> lvim.plugins = {
>   {
>     "simrat39/rust-tools.nvim",
>     config = function()
>       local status_ok, rust_tools = pcall(require, "rust-tools")
>       if not status_ok then
>         return
>       end
> 
>       local opts = {
>         tools = {
>           executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
>           reload_workspace_from_cargo_toml = true,
>           inlay_hints = {
>             auto = true,
>             only_current_line = false,
>             show_parameter_hints = true,
>             parameter_hints_prefix = "<-",
>             other_hints_prefix = "=>",
>             max_len_align = false,
>             max_len_align_padding = 1,
>             right_align = false,
>             right_align_padding = 7,
>             highlight = "Comment",
>           },
>           hover_actions = {
>             --border = {
>             --        { "╭", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╮", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --        { "╯", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╰", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --},
>             auto_focus = true,
>           },
>         },
>         server = {
>           on_attach = require("lvim.lsp").common_on_attach,
>           on_init = require("lvim.lsp").common_on_init,
>           settings = {
>             ["rust-analyzer"] = {
>               checkOnSave = {
>                 command = "clippy"
>               }
>             }
>           },
>         },
>       }
>       --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"
> 
>       --local codelldb_path = extension_path .. "adapter/codelldb"
>       --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"
> 
>       --opts.dap = {
>       --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
>       --}
>       rust_tools.setup(opts)
>     end,
>     ft = { "rust", "rs" },
>   },
> }
> ```
> 
> I did not try to get debugging working at all, hence things being commented out.

iam using this setup and on my termux is working fine.
![Screenshot_20230512-182219_Termux](https://github.com/LunarVim/LunarVim/assets/110718021/d6f52cd9-fec9-45fb-920d-18b75cce6066)

but when i try on my vps its not working also LSP not enable (`:LspInfo` say its enable)
![Screenshot_20230512-182358_Termux](https://github.com/LunarVim/LunarVim/assets/110718021/fbd9a5a6-7c43-479e-a848-bfa4cc840c37)

someone know to fix this issue?

**Comment by lvimuser** [comment]
Debugging codelens works with
```lua
-- https://github.com/simrat39/rust-tools.nvim#a-better-debugging-experience
local path = vim.fn.glob(vim.fn.stdpath("data") .. "/mason/packages/codelldb/extension/") or ""
local codelldb_path = path .. "adapter/codelldb"
local liblldb_path = path .. "lldb/lib/liblldb.so"

if vim.fn.filereadable(codelldb_path) and vim.fn.filereadable(liblldb_path) then
  opts.dap = {
    adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
  }
else
  local msg = "Either codelldb or liblldb is not readable."
    .. "\n codelldb: "
    .. codelldb_path
    .. "\n liblldb: "
    .. liblldb_path
  vim.notify(msg, vim.log.levels.ERROR)
end

rust_tools.setup(opts)
```


not sure if there's a better way. @rlpowell 
Reference:
- https://github.com/abzcoding/lvim/blob/main/lua/user/rust_tools.lua
- https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(via--codelldb)#start-codelldb-automatically

**Comment by abzcoding** [comment]
i've just [updated](https://github.com/abzcoding/lvim/commit/084e278f9f1d62fc421402ff7e13afcf8a563585)  [my settings](https://github.com/abzcoding/lvim/blob/main/lua/user/rust_tools.lua) thanks to @lvimuser 
```lua
local path = vim.fn.glob(vim.fn.stdpath "data" .. "/mason/packages/codelldb/extension/")
    or vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.4/"
  local codelldb_path = path .. "adapter/codelldb"
  local liblldb_path = path .. "lldb/lib/liblldb.so"

  if vim.fn.has "mac" == 1 then
    liblldb_path = path .. "lldb/lib/liblldb.dylib"
  end

  if vim.fn.filereadable(codelldb_path) and vim.fn.filereadable(liblldb_path) then
    opts.dap = {
      adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
    }
  end
  rust_tools.setup(opts)
```

also, you guys might want this
```lua
lvim.builtin.which_key.mappings["ds"] = {
  "<cmd>lua if vim.bo.filetype == 'rust' then vim.cmd[[RustDebuggables]] else require'dap'.continue() end<CR>",
  "Start",
}
```
<img width="895" alt="Screen Shot 2022-09-05 at 12 45 01 PM" src="https://user-images.githubusercontent.com/10992695/188402871-e60ed7c8-1c1e-4f8e-83b7-ac2deee39cdf.png">
<img width="1680" alt="Screen Shot 2022-09-05 at 12 45 30 PM" src="https://user-images.githubusercontent.com/10992695/188402876-c712bef7-1ad2-450d-bb5f-c11a265d1d19.png">

**Comment by badrbouslikhin** [comment]
~~Using the config. proposed in the solution I get the following error: 
`packer.nvim: Error running config for rust-tools.nvim: [string "..."]:0: attempt to index global 'requested_server' (a nil value)`
The plugin seems to be installed though. Any idea?~~

Edit: nevermind, I forgot to `:PackerSync`.

**Comment by matteha** [comment]
Still having trouble getting inlay hints to work unfortunately. 

- I am on LunarVim rolling
- :LspInfo shows rust-analyzer is active, and it highlights errors and so on, but just not inlay hints.

Using the following settings:
```
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })
```
and

```
lvim.plugins = {
  {
    "simrat39/rust-tools.nvim",
    config = function()
      local status_ok, rust_tools = pcall(require, "rust-tools")
      if not status_ok then
        return
      end
      local opts = {
        tools = {
          executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
          reload_workspace_from_cargo_toml = true,
          inlay_hints = {
            auto = true,
            only_current_line = false,
            show_parameter_hints = true,
            parameter_hints_prefix = "<-",
            other_hints_prefix = "=>",
            max_len_align = false,
            max_len_align_padding = 1,
            right_align = false,
            right_align_padding = 7,
            highlight = "Comment",
          },
          hover_actions = {
            --border = {
            --        { "╭", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╮", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --        { "╯", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╰", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --},
            auto_focus = true,
          },
        },
        server = {
          on_attach = require("lvim.lsp").common_on_attach,
          on_init = require("lvim.lsp").common_on_init,
          -- settings = {
          --   ["rust-analyzer"] = {
          --     checkOnSave = {
          --       command = "clippy"
          --     }
          --   }
          -- },
        },
      }
      -- WARNING: I did not try to get debugging working at all, 
      -- hence things being commented out.
      -- See: https://github.com/LunarVim/LunarVim/issues/2894
      --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"
      --local codelldb_path = extension_path .. "adapter/codelldb"
      --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"
      --opts.dap = {
      --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
      --}
      rust_tools.setup(opts)
    end
    -- ft = { "rust", "rs" },
  },
}
```
Also tried

:PackerUpdate
:PackerSync
:LvimCacheReset

and

lua require('rust-tools').inlay_hints_enable()
lua require('rust-tools').inlay_hints_set()

in rust buffers, but to no luck.

**Comment by neevek** [comment]
> It took me like hours? to get this working, because I couldn't find a complete-for-newbs thing that covered the whole thing anywhere, so here's where I ended up in case it helps anybody:
> 
> ```
> 
> lvim.plugins = {
>   {
>     "simrat39/rust-tools.nvim",
>     config = function()
>       local status_ok, rust_tools = pcall(require, "rust-tools")
>       if not status_ok then
>         return
>       end
> 
>       local opts = {
>         tools = {
>           executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
>           reload_workspace_from_cargo_toml = true,
>           inlay_hints = {
>             auto = true,
>             only_current_line = false,
>             show_parameter_hints = true,
>             parameter_hints_prefix = "<-",
>             other_hints_prefix = "=>",
>             max_len_align = false,
>             max_len_align_padding = 1,
>             right_align = false,
>             right_align_padding = 7,
>             highlight = "Comment",
>           },
>           hover_actions = {
>             --border = {
>             --        { "╭", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╮", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --        { "╯", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╰", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --},
>             auto_focus = true,
>           },
>         },
>         server = {
>           on_attach = require("lvim.lsp").common_on_attach,
>           on_init = require("lvim.lsp").common_on_init,
>           settings = {
>             ["rust-analyzer"] = {
>               checkOnSave = {
>                 command = "clippy"
>               }
>             }
>           },
>         },
>       }
>       --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"
> 
>       --local codelldb_path = extension_path .. "adapter/codelldb"
>       --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"
> 
>       --opts.dap = {
>       --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
>       --}
>       rust_tools.setup(opts)
>     end,
>     ft = { "rust", "rs" },
>   },
> }
> ```
> 
> I did not try to get debugging working at all, hence things being commented out.

Thank you so much, it took me two days trying to get inlay hints to work but in vain, until I tried your configurations.

**Comment by rlpowell** [comment]
On Thu, Sep 08, 2022 at 07:13:37AM -0700, matteh wrote:
> Still having trouble getting inlay hints to work unfortunately. 
> 
> - I am on LunarVim rolling
> - :LspInfo shows rust-analyzer is active, and it highlights errors and so on, but just not inlay hints.
> 
> Using the following settings:
> ```
> vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })
> ```
> and
> 
> ```
> lvim.plugins = {
>   {
>     "simrat39/rust-tools.nvim",
>     config = function()
>       local status_ok, rust_tools = pcall(require, "rust-tools")
>       if not status_ok then
>         return
>       end
>       local opts = {
>         tools = {
>           executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
>           reload_workspace_from_cargo_toml = true,
>           inlay_hints = {
>             auto = true,
>             only_current_line = false,
>             show_parameter_hints = true,
>             parameter_hints_prefix = "<-",
>             other_hints_prefix = "=>",
>             max_len_align = false,
>             max_len_align_padding = 1,
>             right_align = false,
>             right_align_padding = 7,
>             highlight = "Comment",
>           },
>           hover_actions = {
>             --border = {
>             --        { "╭", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╮", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --        { "╯", "FloatBorder" },
>             --        { "─", "FloatBorder" },
>             --        { "╰", "FloatBorder" },
>             --        { "│", "FloatBorder" },
>             --},
>             auto_focus = true,
>           },
>         },
>         server = {
>           on_attach = require("lvim.lsp").common_on_attach,
>           on_init = require("lvim.lsp").common_on_init,
>           -- settings = {
>           --   ["rust-analyzer"] = {
>           --     checkOnSave = {
>           --       command = "clippy"
>           --     }
>           --   }
>           -- },
>         },
>       }
>       -- WARNING: I did not try to get debugging working at all, 
>       -- hence things being commented out.
>       -- See: https://github.com/LunarVim/LunarVim/issues/2894
>       --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"
>       --local codelldb_path = extension_path .. "adapter/codelldb"
>       --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"
>       --opts.dap = {
>       --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
>       --}
>       rust_tools.setup(opts)
>     end
>     -- ft = { "rust", "rs" },

To be clear, I'm not a lunarvim or neovim expert at all, I just
shoved a bunch of examples together.

Anyway.

Pretty sure the ft part you commented out is really, really important?

**Comment by ChristianChiarulli** [comment]
```

vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

lvim.plugins = {
  { "catppuccin/nvim", as = "catppuccin" },
  { "lunarvim/darkplus.nvim" },
  { "folke/tokyonight.nvim" },
  {
    "simrat39/rust-tools.nvim",
    config = function()
    end
    -- ft = { "rust", "rs" },
  },
}

local status_ok, rust_tools = pcall(require, "rust-tools")
if not status_ok then
  return
end
local opts = {
  tools = {
    executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
    reload_workspace_from_cargo_toml = true,
    inlay_hints = {
      auto = true,
      only_current_line = false,
      show_parameter_hints = false,
      parameter_hints_prefix = "<-",
      other_hints_prefix = "=>",
      max_len_align = false,
      max_len_align_padding = 1,
      right_align = false,
      right_align_padding = 7,
      highlight = "Comment",
    },
    hover_actions = {
      --border = {
      --        { "╭", "FloatBorder" },
      --        { "─", "FloatBorder" },
      --        { "╮", "FloatBorder" },
      --        { "│", "FloatBorder" },
      --        { "╯", "FloatBorder" },
      --        { "─", "FloatBorder" },
      --        { "╰", "FloatBorder" },
      --        { "│", "FloatBorder" },
      --},
      auto_focus = true,
    },
  },
  server = {
    on_attach = require("lvim.lsp").common_on_attach,
    on_init = require("lvim.lsp").common_on_init,
    -- settings = {
    --   ["rust-analyzer"] = {
    --     checkOnSave = {
    --       command = "clippy"
    --     }
    --   }
    -- },
  },
}
```

This will work just make sure you run the setup outside of the config function.

**Reply by rlpowell** [reply]
That can't be a complete config, because you never use the `opts` variable at all.

**Comment by rlpowell** [comment]
Having spent hours on this again, mostly because I forgot that you have to do :LvimCacheReset when making these sorts of changes -_-;, here's a config with clippy working.  The nature of the `checkOnSave` variable has changed; see https://github.com/simrat39/rust-tools.nvim/issues/208

```
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

lvim.plugins = {
  {
    "simrat39/rust-tools.nvim",
    config = function()
      local status_ok, rust_tools = pcall(require, "rust-tools")
      if not status_ok then
        return
      end

      local opts = {
        tools = {
          executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
          reload_workspace_from_cargo_toml = true,
          inlay_hints = {
            auto = true,
            only_current_line = false,
            show_parameter_hints = true,
            parameter_hints_prefix = "<-",
            other_hints_prefix = "=>",
            max_len_align = false,
            max_len_align_padding = 1,
            right_align = false,
            right_align_padding = 7,
            highlight = "Comment",
          },
          hover_actions = {
            --border = {
            --        { "╭", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╮", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --        { "╯", "FloatBorder" },
            --        { "─", "FloatBorder" },
            --        { "╰", "FloatBorder" },
            --        { "│", "FloatBorder" },
            --},
            auto_focus = true,
          },
        },
        server = {
          on_attach = require("lvim.lsp").common_on_attach,
          on_init = require("lvim.lsp").common_on_init,
          settings = {
            ["rust-analyzer"] = {
              check = {
                command = "clippy",
                extraArgs = { "--all", "--", "-W", "clippy::all" },
              }
            }
          },
        },
      }
      --local extension_path = vim.fn.expand "~/" .. ".vscode/extensions/vadimcn.vscode-lldb-1.7.3/"

      --local codelldb_path = extension_path .. "adapter/codelldb"
      --local liblldb_path = extension_path .. "lldb/lib/liblldb.dylib"

      --opts.dap = {
      --        adapter = require("rust-tools.dap").get_codelldb_adapter(codelldb_path, liblldb_path),
      --}
      rust_tools.setup(opts)
    end,
    ft = { "rust", "rs" },
  },
}
```

**Reply by c02y** [reply]
With this config snippet, every time I open a rust file with lunarvim, there is a line `[LSP] No client with id 1` at the bottom of the vim.

Do you know how to get rid of this message?

**Comment by benediktms** [comment]
Not sure if this is the best place to ask (it's not really a issue with LunarVim I think), but I've been encountering a strange issue, where rust-fmt will no longer format, it seems like it's because rust_analyzer is encountering some sort of error. Has anyone encountered something similar?

Here is the error from the lsp log file:
```log
[ERROR][2023-07-01 01:11:50] .../vim/lsp/rpc.lua:734	"rpc"	"rust-analyzer"	"stderr"	'[ERROR rust_analyzer::dispatch] unknown request: Request { id: RequestId(I32(127)), method: "workspace/executeCommand", params: Object {"title": String("triggerParameterHints"), "command": String("rust-analyzer.triggerParameterHints")} }\n'
[ERROR][2023-07-01 01:14:24] .../vim/lsp/rpc.lua:734	"rpc"	"rust-analyzer"	"stderr"	'[ERROR rust_analyzer::dispatch] unknown request: Request { id: RequestId(I32(688)), method: "workspace/executeCommand", params: Object {"title": String("triggerParameterHints"), "command": String("rust-analyzer.triggerParameterHints")} }\n'
[ERROR][2023-07-01 01:16:06] .../vim/lsp/rpc.lua:734	"rpc"	"rust-analyzer"	"stderr"	"[ERROR hir_ty::mir] Only adt has field    \n[ERROR hir_ty::mir] Only adt has field    \n[ERROR hir_ty::mir] Only adt has field    \n"
```

I'm using rust-tools with a very similar config as above. Also the rust analyzer lsp actually still works. I am getting go to definition, and also clippy. It's just the formatting that doesn't work.

**Comment by mamaraddio** [comment]
> With this config snippet, every time I open a rust file with lunarvim, there is a line `[LSP] No client with id 1` at the bottom of the vim.
> 
> Do you know how to get rid of this message?

I'm having the same problem, also in my case this error appears twice 
![immagine](https://github.com/LunarVim/LunarVim/assets/82162277/de84cb3a-5f99-4a7a-b71a-e049e8a23747)

**Comment by mamaraddio** [comment]
> > With this config snippet, every time I open a rust file with lunarvim, there is a line `[LSP] No client with id 1` at the bottom of the vim.
> > Do you know how to get rid of this message?
> 
> I'm having the same problem, also in my case this error appears twice ![immagine](https://private-user-images.githubusercontent.com/82162277/290821786-de84cb3a-5f99-4a7a-b71a-e049e8a23747.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNjU3MzEsIm5iZiI6MTcwMzA2NTQzMSwicGF0aCI6Ii84MjE2MjI3Ny8yOTA4MjE3ODYtZGU4NGNiM2EtNWY5OS00YTdhLWI3MWEtZTA0OWU4YTIzNzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA5NDM1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5MDY5ZjI1OGJlZjhiY2M1Yjc5YTQxYjM3ZWNkYTQ3NTJkYWE3MGY5NTNkMTJhZGEwMjg3NmE2NmY3NjBkZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8GctoBXOpx7YibZD94a191zRevn0-7tSzpowocnPWA4)

I solved this issue by doing this steps:

Because I'm using not only Rust (to keep my `config.lua` file a bit cleaner), I moved all this configuration (and the Rust keybinding things) to `$HOME/.config/lvim/after/ftplugins/rust.lua` so all this configuration portion is evaluated only if a Rust file is opened

- I removed `rust-analyzer` from automatic configuration server as shown in [starter.lvim](https://github.com/LunarVim/starter.lvim/tree/rust-ide) (`rust-ide` branch)
  In `$HOME/.config/lvim/config.lua`

  ```lua
  vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })
  lvim.plugins = {
    "simrat39/rust-tools.nvim",
    config = true,
    dependencies = {
      {
        "saecki/crates.nvim",
        version = "v0.3.0",
        -- dependencies = { "nvim-lua/plenary.nvim" },  -- I commented it because I alredy have this installed
        config = true
      },
    },
    ft = { "rust", "rs" },
  }

  ```
  This removed one of the two errors
- Safely require `rust-tools` and delayed it's setup
  In `$HOME/.config/lvim/after/ftplugins/rust.lua`

  ```lua
  local ok, rt = pcall(require, "rust-tools")
  if not ok then
	  return
  end
  local ok, crates = pcall(require, "crates")
  if not ok then
	  return
  end

  ...  -- detection of codelldb copied from starter.lvim

  local opts = {
    ... -- your configuration options for rust-tools
  }

  local crates_opts = {
    null_ls = {
      enabled = true,
      name = "crates.nvim",
    },
    popup = {
      border = "rounded",
    },
  }

  local dap_config = function(dap)
    ...  -- DAP options, I copied from starter.lvim
  end

  vim.defer_fn(function() -- defer rust-tools setup to be sure the plugin has loaded
    rt.setup(opts)
    crates.setup(crates_opts)
    lvim.builtin.dap.on_config_done = dap_config
  end, 1000)
  
  ...  -- keybindings
  ```
For the `rust-tools` configuration options refer to [rust-tools.nvim#Configuration](https://github.com/simrat39/rust-tools.nvim#configuration) and [server Configuration Schema](https://github.com/simrat39/rust-tools.nvim/wiki/Server-Configuration-Schema) because some options has changed from the time this issue was opened
